src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/external/StackOverflowClient.java

```diff
@@ -0,0 +1,129 @@
+package com.google.step.YOUR_PROJECT_NAME_HERE.external;
+
+import com.google.step.YOUR_PROJECT_NAME_HERE.data.Card;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.*;
+import java.util.*;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+public final class StackOverflowClient {
+  private static final String SEARCH_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s?"
+          + "order=desc&sort=activity&site=stackoverflow";
+  private static final String QUESTION_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s/answers?"
+          + "order=desc&sort=votes&site=stackoverflow";
+  // this url specify filter to generate answer body
+  private static final String ANSWER_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/answers/%s?order"
+          + "=desc&sort=activity&site=stackoverflow&filter=!9_bDE(fI5";
+  private static final int ID_INDEX = 2;
+
+  public Card search(String so_url) {
+    try {
+      Card card = getQuestion(so_url);
+      card = getAnswerId(card);
+      card = getAnswer(card);
+      return card;
+    } catch (URISyntaxException e) {
+      return new Card();
```

Comment:
I think this function should just return null if a proper card cannot be constructed or if not all fields of the Card can be populated, rather than returning an empty new card. The null Card check can happen in SearchServlet